### ISSUE TITLE
problems installing JAX on a GCP deep learning VM with GPU

### ISSUE THREAD

**Issue by murphyk** [google/jax#9218]
I have created a GCP VM with an A100 GPU and this default image:
c0-deeplearning-common-cu113-v20211219-debian-10
This is cuda_11.3 , CUDNN 8.2 and Debian 10, python 3.7.
I installed JAX thus:
```
pip install --upgrade pip
pip install --upgrade "jax[cuda]" -f https://storage.googleapis.com/jax-releases/jax_releases.html  
```
Inside python 3.7 I type 'import jax' but t I get this error:
```
 version `GLIBCXX_3.4.26' not found
```
According to [this issue, ](https://gitanswer.com/glibcxx-3-4-26-not-found-when-installing-jax-on-gcp-994856532),
I can solve this by first creating a venv and then installing:
```
python -m venv env
source env/bin/activate
pip install --upgrade "jax[cuda]" -f https://storage.googleapis.com/jax-releases/jax_releases.html  
```
This partly works, in that I can now 'import jax' and run it.
However, it fails when I use 'jax.scan':
In particlar, the code snippet below gives this error:
```
2022-01-17 19:46:23.259785: E external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2086] Execution of replica 0 failed: INTERNAL: CustomCall failed: jaxlib/cuda_prng_kernels.cc:30: operation cudaGetLastError() failed: the provided PTX was compiled with an unsupported toolchain.
```
Here is the code:
```
import jax
import jax.numpy as jnp

# sample from a Markov chain
init_dist = jnp.array([0.8, 0.2])
trans_mat = jnp.array([[0.9, 0.1], [0.5, 0.5]])
rng_key = jax.random.PRNGKey(0)
from jax.scipy.special import logit
seq_len = 15

initial_state = jax.random.categorical(rng_key, logits=logit(init_dist), shape=(1,))

def draw_state(prev_state, key):
    logits = logit(trans_mat[:, prev_state])
    state = jax.random.categorical(key, logits=logits.flatten(), shape=(1,))
    return state, state

rng_key, rng_state, rng_obs = jax.random.split(rng_key, 3)
keys = jax.random.split(rng_state, seq_len - 1)

final_state, states = jax.lax.scan(draw_state, initial_state, keys)

print(states)
```

**Comment by zhangqiaorjc** [comment]
This is likely specific to c0-deeplearning-common-cu113-v20211219-debian-10.

@yashk2810 could you help take a look?

**Comment by hawkinsp** [comment]
The `GLIBCXX` version issue was for the last user a `scipy` issue. There's not much we can do other than to drop our scipy dependency or make it optional. That may be possible, we'd have to look into it.

"the provided PTX was compiled with an unsupported toolchain" means that the driver version on the VM is too old for the JAX binary. We may need to build with an older CUDA release. Another option would be for JAX to warn if the CUDA release is too old and hint that the user needs to upgrade their driver.

**Comment by murphyk** [comment]
On Tue, Jan 18, 2022 at 6:12 AM Peter Hawkins ***@***.***>
wrote:

> The GLIBCXX version issue was for the last user a scipy issue. There's
> not much we can do other than to drop our scipy dependency or make it
> optional. That may be possible, we'd have to look into it.
>
> "the provided PTX was compiled with an unsupported toolchain" means that
> the driver version on the VM is too old for the JAX binary. We may need to
> build with an older CUDA release. Another option would be for JAX to warn
> if the CUDA release is too old and hint that the user needs to upgrade
> their driver.
>

How can we do that? Given that this is what every GCP user (who
requesta GPU VM) is going to experience, I think the instructions should be
clear,
otherwise it will just drive users towards other cloud providers.

**Comment by murphyk** [comment]
Any updates?  I verified the problem persists with V100 (not A100) since it's the same machine image.

It seems the same problem was reported in https://github.com/google/jax/issues/6628. I tried that solution (`export XLA_FLAGS="--xla_gpu_cuda_data_dir=/usr/local/cuda/"`) to no avail.

Are you proposing users install CUDA 11.2 themselves just so they can run JAX (following https://cloud.google.com/compute/docs/gpus/install-drivers-gpu)? Since it seems the easy route of using the default '[deep learning VM image' ](https://cloud.google.com/compute/docs/gpus/create-vm-with-gpus#dlvm-image) gives 11.3 which does not work.

BTW the issue with scipy is really about jax, not scipy. If you create a new VM, install jax, start python and type `import jax` you get the scipy GLIBCXX error. By contrast, if you `import scipy` you do not get that error. So I think this error is a warning sign of the deeper problem about incompatible versions.

**Comment by hawkinsp** [comment]
I've never used the "deep learning" images, which is why this has never come up for me. I tend to just install Debian image and install the Python/Nvidia packages myself.

I looked into the GLIBCXX version issue a bit. It's actually a bit more complicated than I thought. It's related to conda.

TL;DR: two workarounds are:
a) `import scipy.signal` before importing JAX. We could add this workaround to jax itself, and I think we should probably just do that for now.
b) make sure Conda's copy of `libstdc++` is first in your library path, e.g.,:
```
$ LD_LIBRARY_PATH=/opt/conda/lib ipython
```

The issue is that the VM image has two copies of libstdc++, one in `/lib` from the OS, and one in `/opt/conda/lib` from Conda. The one in `/lib` is older than the one provided by conda. JAX is happy using either, and it doesn't know anything about the copy from conda. `scipy.signal`, when built from Conda requires the newer copy from conda. (If you used a pip-installed copy of `scipy` it would not have this problem: the manylinux2014 specification forbids.)

Whichever version of libstdc++ you import first in the process is the copy you get. If you import JAX first, then you get the copy from `/lib`, which is too old for the `scipy.signal` package from conda. If you import `scipy.signal` first, then you get Conda's copy, which is newer and everyone is happy.

Note: this isn't really a JAX issue! The same would be true for any pip-installed package that uses C++. For example, you can reproduce this with TensorFlow:
```
$ pip install tensorflow
$ ipython
Python 3.7.12 | packaged by conda-forge | (default, Oct 26 2021, 06:08:53)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.30.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import tensorflow

In [2]: import scipy.signal
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-2-221f770a1289> in <module>
----> 1 import scipy.signal

/opt/conda/lib/python3.7/site-packages/scipy/signal/__init__.py in <module>
    292 from . import sigtools, windows
    293 from .waveforms import *
--> 294 from ._max_len_seq import max_len_seq
    295 from ._upfirdn import upfirdn
    296

/opt/conda/lib/python3.7/site-packages/scipy/signal/_max_len_seq.py in <module>
      6 import numpy as np
      7
----> 8 from ._max_len_seq_inner import _max_len_seq_inner
      9
     10 __all__ = ['max_len_seq']

ImportError: /lib/x86_64-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.26' not found (required by /opt/conda/lib/python3.7/site-packages/scipy/signal/_max_len_seq_inner.cpython-37m-x86_64-linux-gnu.so)
```

or with PyTorch:
```
$ pip install torch
$ ipython
Python 3.7.12 | packaged by conda-forge | (default, Oct 26 2021, 06:08:53)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.30.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import torch

In [2]: import scipy.signal
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-2-221f770a1289> in <module>
----> 1 import scipy.signal

/opt/conda/lib/python3.7/site-packages/scipy/signal/__init__.py in <module>
    292 from . import sigtools, windows
    293 from .waveforms import *
--> 294 from ._max_len_seq import max_len_seq
    295 from ._upfirdn import upfirdn
    296

/opt/conda/lib/python3.7/site-packages/scipy/signal/_max_len_seq.py in <module>
      6 import numpy as np
      7
----> 8 from ._max_len_seq_inner import _max_len_seq_inner
      9
     10 __all__ = ['max_len_seq']

ImportError: /lib/x86_64-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.26' not found (required by /opt/conda/lib/python3.7/site-packages/scipy/signal/_max_len_seq_inner.cpython-37m-x86_64-linux-gnu.so)
```

I can add the workaround of importing `scipy.signal` to JAX before we import any of JAX's own libraries, effectively replicating that workaround. But I'd argue it's really a Conda bug, although I'm not quite sure how to fix it.

Another possible workaround would be if JAX provided conda packages and you installed JAX via Conda.

I'll look more at the CUDA versioning issue later.

**Comment by hawkinsp** [comment]
I looked into the CUDA version problem. The issue is that the driver installed by the deep learning VM (NB. the *driver* version, not the CUDA version: the CUDA version is fine) is too old for the JAX wheels, in the case that you are using a GPU (like an A100) for which we don't explicitly compile SASS code.

Good news: we added A100 to the list of GPUs we support for the next jaxlib release (0.1.76, which hopefully we will release soon).

(I was unable to reproduce this problem on a V100.)

I think we may also need to change at least one of our wheels to be built with a lower driver version requirement for folks with older drivers to use. I'm not sure why the deep learning VM has such an old driver version. I'll ask them if it can be upgraded.

Building jaxlib from source on the VM would have solved this problem also.

**Comment by hawkinsp** [comment]
I verified that upgrading the driver to `470.82.01` on the VM fixed the problem (well, it takes a few minutes to start JAX the first time because it has to build a bunch of PTX code for the A100 GPU, but the next jaxlib release will fix that).

**Comment by murphyk** [comment]
Great news, thanks! When will the new jaxlib release come out?

**Comment by murphyk** [comment]
I see that[ jaxlib 0.1.76 is released](https://jax.readthedocs.io/en/latest/changelog.html#jaxlib-0-1-76-jan-27-2022)!
On my A100 machine, I uninstalled jax and jaxlib and then ran 
```
pip install --upgrade pip
pip install --upgrade "jax[cuda]" -f https://storage.googleapis.com/jax-releases/jax_releases.html  
```
but it still uses jaxlib-0.1.75 (see below).
I think you need to update https://storage.googleapis.com/jax-releases/jax_releases.html?


```
 pip install --upgrade "jax[cuda]" -f https://storage.googleapis.com/jax-releases/jax_releases.html  
Looking in links: https://storage.googleapis.com/jax-releases/jax_releases.html
Requirement already satisfied: jax[cuda] in /opt/conda/lib/python3.7/site-packages (0.2.27)
Requirement already satisfied: absl-py in /opt/conda/lib/python3.7/site-packages (from jax[cuda]) (1.0.0)
Requirement already satisfied: opt-einsum in /opt/conda/lib/python3.7/site-packages (from jax[cuda]) (3.3.0)
Requirement already satisfied: scipy>=1.2.1 in /opt/conda/lib/python3.7/site-packages (from jax[cuda]) (1.7.3)
Requirement already satisfied: numpy>=1.19 in /opt/conda/lib/python3.7/site-packages (from jax[cuda]) (1.19.5)
Requirement already satisfied: typing-extensions in /opt/conda/lib/python3.7/site-packages (from jax[cuda]) (4.0.1)
Collecting jaxlib==0.1.75+cuda11.cudnn82
  Using cached https://storage.googleapis.com/jax-releases/cuda11/jaxlib-0.1.75%2Bcuda11.cudnn82-cp37-none-manylinux2010_x86_64.whl (149.4 MB)
Requirement already satisfied: flatbuffers<3.0,>=1.12 in /opt/conda/lib/python3.7/site-packages (from jaxlib==0.1.75+cuda11.cudnn82->jax[cuda]) (2.0)
Requirement already satisfied: six in ./.local/lib/python3.7/site-packages (from absl-py->jax[cuda]) (1.16.0)
Installing collected packages: jaxlib
Successfully installed jaxlib-0.1.75+cuda11.cudnn82
```

**Comment by hawkinsp** [comment]
It's... not quite yet released. We found one last bug in our presubmit testing.

But if you manually download e.g., https://storage.googleapis.com/jax-releases/cuda11/jaxlib-0.1.76+cuda11.cudnn805-cp39-none-manylinux2010_x86_64.whl  substituting the python version as appropriate it may work for you. We may need to revise it before publishing it though.

**Comment by hawkinsp** [comment]
We pushed the new wheels, so this should now work. I verified your example works on my DL GPU image on A100 (modulo the "you should use a virtual env or import scipy.signal", problem, for which workarounds are discussed above). Hope that helps!

**Comment by murphyk** [comment]
I made a venv, and  tried
```
pip install https://storage.googleapis.com/jax-releases/cuda11/jaxlib-0.1.76+cuda11.cudnn82-cp37-none-manylinux2010_x86_64.whl
```
In python I have the latest version of jax and jaxlib
```
>>> import jaxlib
>>> jaxlib.__version__
'0.1.76'
>>> import jax
ja>>> jax.__version__
'0.2.27'
```
Most things seem to work, except for RNG:
In particular, this fails:
```
 import jax; import jax.numpy as jnp; from jax.scipy.special import logit; init_dist = jnp.array([0.8, 0.2]); rng_key = jax.random.PRNGKey(0);initial_state = jax.random.categorical(rng_key, logits=logit(init_dist), shape=(1,))
```
with this error
```
 File "/home/kpmurphy_google_com/env/lib/python3.7/site-packages/jax/_src/dispatch.py", line 444, in _execute_compiled
    out_bufs = compiled.execute(input_bufs)
jax._src.traceback_util.UnfilteredStackTrace: RuntimeError: INTERNAL: CustomCall failed: jaxlib/cuda_prng_kernels.cc:30: operation cudaGetLastError() failed: the provided PTX was compiled with an unsupported toolchain.

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/kpmurphy_google_com/env/lib/python3.7/site-packages/jax/_src/random.py", line 1212, in categorical
    gumbel(key, sample_shape + logits.shape, logits.dtype) +
  File "/home/kpmurphy_google_com/env/lib/python3.7/site-packages/jax/_src/random.py", line 1170, in gumbel
    return _gumbel(key, shape, dtype)
  File "/home/kpmurphy_google_com/env/lib/python3.7/site-packages/jax/_src/dispatch.py", line 444, in _execute_compiled
    out_bufs = compiled.execute(input_bufs)
RuntimeError: INTERNAL: CustomCall failed: jaxlib/cuda_prng_kernels.cc:30: operation cudaGetLastError() failed: the provided PTX was compiled with an unsupported toolchain.
```

**Comment by hardmaru** [comment]
I'm also having this issue on GCP's VMs with A100s.

Would like to note that for some reason, every GCP VM comes default with the 460 drivers (even NVIDIA's official images for deep learning and HPC).

Will see if I can find a solution later.

**Comment by murphyk** [comment]
Quick update. I created a new V100 VM with Deep Learning image but chose the pytorch image 

![image](https://user-images.githubusercontent.com/4632336/158315894-2628cece-b5fd-4d35-9d15-e6e78ac2bfcf.png)

This has these specs
```
nvidia-smi
NVIDIA-SMI 460.73.01    Driver Version: 460.73.01    CUDA Version: 11.2  

cat /usr/local/cuda/include/cudnn_version.h | grep CUDNN_MAJOR -A 2

define CUDNN_MAJOR 8
#define CUDNN_MINOR 0
#define CUDNN_PATCHLEVEL 5
```

I could install jax using 
```
pip install jax[cuda11_cudnn805] -f https://storage.googleapis.com/jax-releases/jax_releases.html
```
This runs without errors (including `import scipy.signal`), but I get warnings when I use RNG:
```
external/org_tensorflow/tensorflow/stream_executor/gpu/asm_compiler.cc:111] *** WARNING *** You are using ptxas 11.0.221, which is older than 11.1. ptxas before 11.1 is known to miscompile XLA code, leading to incorrect results or invalid-address errors.
```

**Comment by murphyk** [comment]
I also created a VM using cuda 11.2
```
[c0-deeplearning-common-cu113-v20220227-debian-10](https://pantheon.corp.google.com/compute/imagesDetail/projects/ml-images/global/images/c0-deeplearning-common-cu113-v20220227-debian-10?project=kpmurphy-gcp2)

 NVIDIA-SMI 460.73.01    Driver Version: 460.73.01    CUDA Version: 11.2 
```
which now won't even run the RNG code:
```
RuntimeError: INTERNAL: CustomCall failed: jaxlib/cuda_prng_kernels.cc:30: operation cudaGetLastError() failed: the provided PTX was compiled with an unsupported toolchain.
```

So the bottom line is: it seems very unclear how to run JAX on GCP unless you are willing to install CUDA from scratch,
which seems hard.
In particular,  according to [go/step-by-step-gcp](http://go/step-by-step-gcp), the steps are:

Manually install NVIDIA driver 470.82.0[1](https://github.com/ashutoshIITK/install_cuda_cudnn_ubuntu_20), CUDA 11.4.4, and CuDNN 8.[2](https://medium.com/@anarmammadli/how-to-install-cuda-11-4-on-ubuntu-18-04-or-20-04-63f3dee2099).4 (Here are two blog posts about how one might go about doing this: 1, 2). Note that you will need to register for a NVIDIA’s developer account to download the CuDNN tarball. You can download it on your local machine, and upload that file using the web ssh interface.

https://github.com/ashutoshIITK/install_cuda_cudnn_ubuntu_20

**Comment by sayakpaul** [comment]
I think there's still problem.

`nvidia-smi`

![image](https://user-images.githubusercontent.com/22957388/179160650-be0ab22f-5de4-488e-a5de-f16072f25aec.png)


`cat /usr/local/cuda/include/cudnn_version.h | grep CUDNN_MAJOR -A 2`

```
#define CUDNN_MAJOR 8
#define CUDNN_MINOR 0
#define CUDNN_PATCHLEVEL 5
--
#define CUDNN_VERSION (CUDNN_MAJOR * 1000 + CUDNN_MINOR * 100 + CUDNN_PATCHLEVEL)

#endif /* CUDNN_VERSION_H */
```

`nvcc --version`

```
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2020 NVIDIA Corporation
Built on Wed_Jul_22_19:09:09_PDT_2020
Cuda compilation tools, release 11.0, V11.0.221
Build cuda_11.0_bu.TC445_37.28845127_0
```

I tried with `pip install "jax[cuda11_cudnn805]" -f https://storage.googleapis.com/jax-releases/jax_cuda_releases.html`. It did work but it complains:

```
You may not need to update to CUDA 11.1; cherry-picking the ptxas binary is often sufficient.
2022-07-15 06:05:36.687886: W external/org_tensorflow/tensorflow/stream_executor/gpu/asm_compiler.cc:111] *** WARNING *** You are using ptxas 11.0.221, which is older than 11.1. ptxas before 11.1 is known to miscompile XLA code, leading to incorrect results or invalid-address errors.
```

Any workarounds?

**Comment by hawkinsp** [comment]
The CUDA driver version incompatibility problem should be fixed in the next jaxlib release. JAX will automatically fall back to not using parallel compilation if the NVIDIA driver is too old.

Unfortunately we had to revert the workaround for `version `GLIBCXX_3.4.26' not found` because the workaround was to import `scipy` ourselves, but that turns out to be too slow to do every time jax is imported. If you still see that problem, I recommend one of the workarounds above. Note that `jax` is also available via `conda-forge` (https://github.com/google/jax#conda-installation) and using the `conda` installation of JAX will not have this issue.

@sayakpaul the issue is you have CUDA 11.0 installed. JAX doesn't support CUDA 11.0. Install a newer CUDA.
